File: Problems/Amazon Locker/Java/Locker.java
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class Locker {
    private List<Compartment> compartments;
    private Map<String, AccessToken> accessTokenMap;

    public Locker(List<Compartment> compartments) {
        this.compartments = compartments;
        this.accessTokenMap = new HashMap<>();
    }

    public String depositPackage(Size size) {
        Compartment compartment = findUnOccupiedCompartment(size);
        if(compartment == null) {
            throw new RuntimeException("no available compartment of size " + size);
        }

        compartment.open();
        compartment.markOccupied();

        AccessToken accessToken = generateAccessToken(compartment);
        accessTokenMap.put(accessToken.getCode(), accessToken);

        return accessToken.getCode();
    }

    public void pickup(String accessCode) {
        if("".equals(accessCode)) {
            throw new IllegalArgumentException("invalid accessCode!");
        }

        AccessToken accessToken = accessTokenMap.get(accessCode);
        if(accessToken.isExpired()) {
            throw new RuntimeException("access token has expired!");
        }

        Compartment compartment = accessToken.getCompartment();
        compartment.open();

        clearDeposit(accessToken);
    }

    private void clearDeposit(AccessToken accessToken) {
        Compartment compartment = accessToken.getCompartment();
        compartment.markFree();
        accessTokenMap.remove(accessToken.getCode());
    }

    public void openExpiredCompartment() {
        for(AccessToken accessToken : accessTokenMap.values()) {
            if(accessToken.isExpired()) {
                accessToken.getCompartment().open();
            }
        }
    }

    private Compartment findUnOccupiedCompartment(Size size) {
        for(Compartment compartment : compartments) {
            if(!compartment.isOccupied()) {
                return compartment;
            }
        }
        return null;
    }

    private AccessToken generateAccessToken(Compartment compartment) {
        Random random = new Random();
        String accessCode = String.format("%06d", random.nextInt(1000000));
        LocalDateTime expiration = LocalDateTime.now().plusDays(7);
        return new AccessToken(accessCode, expiration, compartment);
    }
}

File: Problems/Amazon Locker/Go/locker.go
package main

import (
	"fmt"
	"math/rand/v2"
	"time"
)

type Locker struct {
	compartments   []*Compartment
	accessTokenMap map[string]*AccessToken
}

func NewLocker(compartments []*Compartment) *Locker {
	return &Locker{
		compartments:   compartments,
		accessTokenMap: make(map[string]*AccessToken),
	}
}

func (l *Locker) DepositPackage(size Size) (string, error) {
	compartment := l.findUnOccupiedCompartmentBySize(size)
	if compartment == nil {
		return "", fmt.Errorf("no available compartment of size %s", size)
	}

	compartment.Open()
	compartment.MarkOccupied()
	accessToken := l.generateAccessToken(compartment)
	l.accessTokenMap[accessToken.GetCode()] = accessToken

	return accessToken.GetCode(), nil
}

func (l *Locker) Pickup(accessCode string) error {
	accessToken, ok := l.accessTokenMap[accessCode]
	if !ok {
		return fmt.Errorf("invalid access token code")
	}
	if accessToken.IsExpired() {
		return fmt.Errorf("access token has expired")
	}

	compartment := accessToken.GetCompartment()
	compartment.Open()
	l.clearDeposit(accessToken)

	return nil
}

func (l *Locker) OpenExpiredCompartment() {
	for _, accessToken := range l.accessTokenMap {
		if accessToken.IsExpired() {
			accessToken.GetCompartment().Open()
		}
	}
}

func (l *Locker) findUnOccupiedCompartmentBySize(size Size) *Compartment {
	for _, compartment := range l.compartments {
		if !compartment.IsOccupied() && compartment.GetSize() == size {
			return compartment
		}
	}
	return nil
}

func (l *Locker) generateAccessToken(compartment *Compartment) *AccessToken {
	code := fmt.Sprintf("%06d", rand.IntN(1000000))
	expiration := time.Now().Add(7 * 24 * time.Hour)
	return NewAccessToken(code, expiration, compartment)
}

func (l *Locker) clearDeposit(token *AccessToken) {
	compartment := token.GetCompartment()
	compartment.MarkFree()
	delete(l.accessTokenMap, token.GetCode())
}
